untrusted/quote_verify/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+// the Quote Signature Data Structure. Table 4 of
+// https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
+const QUOTING_ENCLAVE_SIGNATURE_START: usize = QUOTING_ENCLAVE_REPORT_START + ENCLAVE_REPORT_SIZE;
+
+// ASN.1 Tag for an integer
+const ASN1_INTEGER: u8 = 2;
+// ASN.1 Tag for a sequence
+const ASN1_SEQUENCE: u8 = 48;
+
+/// A quote for DCAP attestation
+pub struct Quote {
+    bytes: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Unable to load the Certificate with mbedtls
+    Certificate(mbedtls::Error),
+    /// Failure to parse the Pem files from the quote data
+    PemParsing(PemError),
+    /// Failure to verify a Signature
+    Signature(mbedtls::Error),
+}
+
+impl From<PemError> for Error {
+    fn from(src: PemError) -> Self {
+        Self::PemParsing(src)
+    }
+}
+
+impl Quote {
+    /// Returns a [Quote] created from the provided `bytes`.
+    ///
+    /// # Arguments
+    ///
+    /// * `bytes` the bytes of the quote as defined in
+    ///     https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        Quote {
+            bytes: bytes.to_vec(),
+        }
+    }
+
+    /// Verify the quoting enclave within the report.
+    pub fn verify_quoting_enclave_report(&self) -> Result<(), Error> {
+        let signature = self.get_ans1_signature();
+        let report = self.get_quoting_enclave_report();
+        let hash = Sha256::digest(report);
+        let mut cert = self.get_pck_certificate()?;
+        cert.public_key_mut()
+            .verify(HashType::Sha256, &hash, &signature)
+            .map_err(Error::Signature)
+    }
+
+    /// Returns the PCK certificate that was used for signing the quoting
+    /// enclave report.
+    /// Note: This certificate is assumed to be valid.
+    fn get_pck_certificate(&self) -> Result<MbedtlsBox<Certificate>, Error> {
+        //TODO this should only be looking at the `Certification Data`
+        // of the `QE Certification Data`, Table 9 of
+        // https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf
+        // However the pem crate walks over the other data nicely for initial development
+        let pem = pem::parse(self.bytes.as_slice())?;
+        Certificate::from_pem(&pem.contents).map_err(Error::Certificate)
+    }
+
+    /// Returns the quoting enclave report from the overall quote.
+    fn get_quoting_enclave_report(&self) -> &[u8] {
+        &self.bytes
+            [QUOTING_ENCLAVE_REPORT_START..QUOTING_ENCLAVE_REPORT_START + ENCLAVE_REPORT_SIZE]
+    }
+
+    /// Returns the ASN.1 version of the signature.
+    /// mbedtls wants an ASN.1 version of the signature, while the raw quote
+    /// format has only the r and s values of the ECDSA signature
+    fn get_ans1_signature(&self) -> Vec<u8> {
```

Comment:
Typo, should be `asn1`

Reply:
good catch!  I kept googling ans1 and was getting a bunch of ansi stuff :)